src/pages/home/report/withReportAndPrivateNotesOrNotFound.tsx

```diff
@@ -1,88 +1,63 @@
-     * That way, if a ref is passed to a component wrapped in the HOC, the ref is a reference to the wrapped component, not the HOC. */
-    forwardedRef: PropTypes.func,
-
-    /** The report currently being looked at */
-    report: reportPropTypes,
-
-    /** Information about the network */
-    network: networkPropTypes.isRequired,
-
+type WithReportAndPrivateNotesOrNotFoundOnyxProps = {
     /** Session of currently logged in user */
-    session: PropTypes.shape({
-        /** accountID of currently logged in user */
-        accountID: PropTypes.number,
-    }),
-
-    route: PropTypes.shape({
-        /** Params from the URL path */
-        params: PropTypes.shape({
-            /** reportID and accountID passed via route: /r/:reportID/notes/:accountID */
-            reportID: PropTypes.string,
-            accountID: PropTypes.string,
-        }),
-    }).isRequired,
+    session: OnyxEntry<OnyxTypes.Session>;
 };
 
-const defaultProps = {
-    forwardedRef: () => {},
-    report: {},
-    session: {
-        accountID: null,
-    },
-};
+type WithReportAndPrivateNotesOrNotFoundProps = WithReportAndPrivateNotesOrNotFoundOnyxProps & WithReportOrNotFoundProps & {route: RouteProp<{params: {accountID?: string}}>};
 
-export default function (pageTitle) {
+export default function <TProps extends WithReportAndPrivateNotesOrNotFoundProps, TRef>(pageTitle: TranslationPaths) {
     // eslint-disable-next-line rulesdir/no-negated-variables
-    return (WrappedComponent) => {
+    return (WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>) => {
         // eslint-disable-next-line rulesdir/no-negated-variables
-        function WithReportAndPrivateNotesOrNotFound({forwardedRef, ...props}) {
+        function WithReportAndPrivateNotesOrNotFound(props: TProps & {forwardedRef: ForwardedRef<TRef>}) {
             const {translate} = useLocalize();
-            const {route, report, network, session} = props;
-            const accountID = route.params.accountID;
-            const isPrivateNotesFetchTriggered = !_.isUndefined(report.isLoadingPrivateNotes);
+            const network = useNetwork();
+            const {route, report, session, forwardedRef} = props;
+            const accountID = route.params?.accountID;
+            const isPrivateNotesFetchTriggered = report?.isLoadingPrivateNotes !== undefined;
             const prevIsOffline = usePrevious(network.isOffline);
             const isReconnecting = prevIsOffline && !network.isOffline;
-            const isOtherUserNote = accountID && Number(session.accountID) !== Number(accountID);
+            const isOtherUserNote = accountID && Number(session?.accountID) !== Number(accountID);
             const isPrivateNotesFetchFinished = isPrivateNotesFetchTriggered && !report.isLoadingPrivateNotes;
-            const isPrivateNotesEmpty = accountID ? _.has(lodashGet(report, ['privateNotes', accountID, 'note'], '')) : _.isEmpty(report.privateNotes);
```

Comment:
Do you understand what was happening here?

```ts
_.has(lodashGet(report, ['privateNotes', accountID, 'note'], ''))
```

What does this mean?

Reply:
Doesn't seem to be consistent with [the docs](https://lodash.com/docs/4.17.15#has), though.

Reply:
I took a closer look into the docs and it seems that the previous expression would always be evaluated as false, which doesn't make sense

Reply:
@cubuspl42 I'll reply to your last feedback here since all this discussion is the same.

Let's take a look into `const shouldShowFullScreenLoadingIndicator = !isPrivateNotesFetchFinished || (isPrivateNotesEmpty && (report.isLoadingPrivateNotes || !isOtherUserNote));`.

If `isPrivateNotesFetchFinished` is true, then `const shouldShowFullScreenLoadingIndicator = isPrivateNotesEmpty && (report.isLoadingPrivateNotes || !isOtherUserNote);`

Like I said here, previously `isPrivateNotesEmpty` would always be false if the `accountID` was present, because `_.has(lodashGet(report, ['privateNotes', accountID, 'note'], ''))` was badly formed because of the missing second argument according to `lodash` docs.

Then, when accountID is is present, `const shouldShowFullScreenLoadingIndicator = false`

Now let's take a look at the other possibility of accountID not being present:

- `const isOtherUserNote = accountID && Number(session.accountID) !== Number(accountID);` => `const isOtherUserNote = false`
- `const isPrivateNotesEmpty = accountID ? _.has(lodashGet(report, ['privateNotes', accountID, 'note'], '')) : _.isEmpty(report.privateNotes);` => `const isPrivateNotesEmpty =  _.isEmpty(report.privateNotes);`

Now, if `isPrivateNotesFetchFinished` and `isPrivateNotesEmpty` are true, then `const shouldShowFullScreenLoadingIndicator = report.isLoadingPrivateNotes || !isOtherUserNote;`

Since  `const isPrivateNotesFetchFinished = isPrivateNotesFetchTriggered && !report.isLoadingPrivateNotes;`, if isPrivateNotesFetchFinished is true then it implies that `report.isLoadingPrivateNotes` is false, so `const shouldShowFullScreenLoadingIndicator = !isOtherUserNote;`

Now, since `isOtherUserNote` is false, then `const shouldShowFullScreenLoadingIndicator = true` and the infinite loader is shown instead of the not found page.

From my understanding, `const shouldShowFullScreenLoadingIndicator = !isPrivateNotesFetchFinished || (isPrivateNotesEmpty && (report.isLoadingPrivateNotes || !isOtherUserNote));` is the same as `const shouldShowFullScreenLoadingIndicator = !isPrivateNotesFetchFinished || (isPrivateNotesEmpty && !isOtherUserNote);`, but this implies that an infinite loader will show up when no private notes exist for a report because the loading page takes precedence to the not found page:

```
if (shouldShowFullScreenLoadingIndicator) {
                return <LoadingPage title={translate(pageTitle)} />;
            }

            if (shouldShowNotFoundPage) {
                return <NotFoundPage />;
            }
```

That's why I removed the `(isPrivateNotesEmpty && !isOtherUserNote);` thus the not found page is shown instead when no accountID is provided like it's defined on 

```
            const shouldShowNotFoundPage = useMemo(() => {
                // Show not found view if the report is archived, or if the note is not of current user.
                if (ReportUtils.isArchivedRoom(report) || (accountID && Number(session.accountID) !== Number(accountID))) {
                    return true;
                }

                // Don't show not found view if the notes are still loading, or if the notes are non-empty.
                if (shouldShowFullScreenLoadingIndicator || !isPrivateNotesEmpty || isReconnecting) {
                    return false;
                }

                // As notes being empty and not loading is a valid case, show not found view only in offline mode.
                return network.isOffline;
            }, [report, network.isOffline, accountID, session.accountID, isPrivateNotesEmpty, shouldShowFullScreenLoadingIndicator, isReconnecting]);
```

Let me now if my explanation clarified your questions, otherwise let me know if you have any other questions

Reply:
I'm in two minds here...

On the one hand, I agree that the TypeScript migration is sometimes an opportunity to fix some obvious issues of the surrounding code. I believe it should be done when there's _some_ typing benefit of the changed behavior.

On the other hand, the code you submitted for review was non-functional **twice**, while the original code worked. By simple reasoning, it should keep working after migrating to TypeScript, as TypeScript is _mostly_ just a statically typed JavaScript superset.

Although my heart is with you, as I'm also an engineer, for any business, the priorities are like this:

Working code >> Clean code > Unclean code

Following this reasoning, I believe that TypeScript migrations should do only minimal behavior-changing refactoring.